### PR TITLE
Tutorial: Create your First App with Gutenberg Data

### PR DESCRIPTION
**Pull request by adamziel**
## Description

This PR proposes a tutorial gently introducing the user to Gutenberg Data's concepts through building a small application.

It is a spin-off of https://github.com/WordPress/gutenberg/pull/37615 dedicated for new users. I would like to follow up with one or two more advanced materials.

For now I drafted the Readme, setup, and the part where we build the list view for WordPress pages. I'd like to stop and collect feedback before writing the rest of it. Perhaps we could merge this tutorial part-by-part instead of all at once.

* [Rich-text preview.](https://github.com/WordPress/gutenberg/tree/doc/first-app-with-gutenberg-data/docs/how-to-guides/data-basics)
* [Related gutenberg-examples PR](https://github.com/WordPress/gutenberg-examples/pull/191)

Kudos to @getdave @bph @ironprogrammer @ryanwelcher @mkaz @youknowriad @gziolo @talldan @tellthemachines @draganescu for their feedback so far!

cc @draganescu @gziolo @youknowriad @talldan @annezazu @noisysocks

### CODE REVIEW

File: docs/how-to-guides/data-basics/1-setup.md
Comment (youknowriad) [inline docs/how-to-guides/data-basics/1-setup.md]: should we explain what each one of the dependencies provide for our example?

**Review by youknowriad — COMMENTED** [review]
I like this tutorial a lot. I didn't read fully but somethings standout to me: the repeating of `wp.element.createElement` is not great and aside JSX, it's seem that you're using ESnext in the examples. It makes me wonder whether it's better to just use JSX in the examples and have a note at the beginning or the end that links to the build setup config tutorial.

**Comment by Mamaduka** [comment]
Stellar work, @adamziel 🥇 

Will this tutorial be part of the Gutenberg examples repo?

I usually find it helpful to see finished example projects and how all pieces come together.

P.S. I assumed that you need to wrap your app in `EntityProvider` when working with entities; I see that this isn't a hard requirement.

**Comment by gziolo** [comment]
> Will this tutorial be part of the Gutenberg examples repo?

That would be great 👍

> It makes me wonder whether it's better to just use JSX in the examples and have a note at the beginning or the end that links to the build setup config tutorial.

I agree that tutorials should promote JSX and ES Modules in the first place. It would be good to double check whether they still provide tye version without JSX.

The first exposure to coding tutorials is here: https://github.com/WordPress/gutenberg/tree/trunk/docs/getting-started/create-block

I haven’t checked the tutorial yet but I love the idea ❤️

**Comment by adamziel** [comment]
I just realized it's not obvious how to review the rendered markdown and not just the code: [Rendered markdown](https://github.com/WordPress/gutenberg/tree/doc/first-app-with-gutenberg-data/docs/how-to-guides/data-basics)

~~Here's a few links to rich-text previews:~~

* ~~[Readme](https://github.com/WordPress/gutenberg/pull/38250/files?short_path=78f6d8e#diff-78f6d8e3bf1d87a05f55f9516aaa5494f8d91c1e7b98e541820b7a01cb67f0f8)~~
* ~~[Setup](https://github.com/WordPress/gutenberg/pull/38250/files?short_path=6371135#diff-63711350cc2580527dbc8d1cc63ec912faa1d3caaef5e3181567675446efcaee)~~
* ~~[Building a list of pages](https://github.com/WordPress/gutenberg/pull/38250/files?short_path=b430cc0#diff-b430cc03c3beb60fedd5c77bc6c44789129a8b491d4bbfe77fbc61829c3586f2)~~

~~Unfortunately the screenshots don't render. If anyone knows how to work around that, please tell us :D~~

File: docs/how-to-guides/data-basics/1-setup.md
Comment (adamziel) [reply docs/how-to-guides/data-basics/1-setup.md]: Good point! I added a short comment for each

**Comment by adamziel** [comment]
I improved the screenshots, polished the grammar, and added finished examples. Here's the [rich-text preview](https://github.com/WordPress/gutenberg/tree/doc/first-app-with-gutenberg-data/docs/how-to-guides/data-basics) of the tutorial so far.

Can y'all see anything else that can be improved before merging the first part of it?

> Will this tutorial be part of the Gutenberg examples repo?

@Mamaduka I would love that! And I think we're pretty close to making it happen by merging this PR :-) 

> I like this tutorial a lot. I didn't read fully but somethings standout to me: the repeating of wp.element.createElement is not great and aside JSX, it's seem that you're using ESnext in the examples. It makes me wonder whether it's better to just use JSX in the examples and have a note at the beginning or the end that links to the build setup config tutorial.

@youknowriad I went for JSX-first examples with a hidden compiled snippet. I think that's the best of both worlds, as in we have readability and convenience at the same time.

As for using ESnext in the examples – many features are widely supported by browsers so I think that's fine as long as it's not sprinkled with JSX.

> I agree that tutorials should promote JSX and ES Modules in the first place.

@gziolo I went for `wp.data` in this tutorial as I wasn't sure how to access global stores via importing ES modules. Is it possible?

> I usually find it helpful to see finished example projects and how all pieces come together.

It's now a part of this PR! :-)

---

@getdave I know you had the following notes:

> Introduce the convenience entity selectors and resolvers such as getMenus

Pages unfortunately don't have `getPages` selectors and I am not sure how to make a clear distinction between hard-coded entities and different types pages. Would you like to propose something?

> Be ultra clear that the resolver needs the exact same arguments as the selector in order to work

How would this look like? I didn't know how to introduce this topic as we're not building a new resolver at this point, only calling a selector that happens to have a related resolver.

File: docs/how-to-guides/data-basics/1-setup.md
Comment (gziolo) [inline docs/how-to-guides/data-basics/1-setup.md]: I think you can skip this whole paragraph or link to some existing page that covers the same topic. 

A good reference would be also a tutorial that @getdave contributed for the standalone block editor:

https://github.com/WordPress/gutenberg/blob/trunk/docs/how-to-guides/platform/custom-block-editor/tutorial.md

Those are advanced tutorials and they should be using the most efficient syntax so in my opinion the build step is essential here.

Comment (adamziel) [reply docs/how-to-guides/data-basics/1-setup.md]: I agree in principle, but I think of this tutorial as more of a first step towards understanding advanced Gutenberg concepts. It can benefit from a build tool but I don't think it should _require_ one. I'm thinking of myself here – I'm pretty impatient and I would instinctively look for a different tutorial with less pre-requisites. 

That being said, let's see if I can make this section more succinct.

File: docs/how-to-guides/data-basics/1-setup.md
Comment (Mamaduka) [inline docs/how-to-guides/data-basics/1-setup.md]:
```suggestion
You don’t need a build tool to complete this tutorial, but using one will make your code look nicer.
```

Typo.

File: docs/how-to-guides/data-basics/1-setup.md
Comment (Mamaduka) [inline docs/how-to-guides/data-basics/1-setup.md]:
```suggestion
* app.js – for our JavaScript application
```

Based on filename usage in the tutorial.

File: docs/how-to-guides/data-basics/1-setup.md
Comment (gziolo) [reply docs/how-to-guides/data-basics/1-setup.md]: That's why we always try to pair the tutorial with https://github.com/WordPress/gutenberg-examples so you don't have to write a single line of code to run it 😄

Comment (adamziel) [reply docs/how-to-guides/data-basics/1-setup.md]: Ooh that's great, thank you for sharing! I'll look into that

**Comment by ryanwelcher** [comment]
First, this is a fantastic resource and I'm so happy you've taken the time to put it together. I think it'll be a great addition to https://github.com/WordPress/gutenberg-examples. 

My only concern, which has been shared by others already, is we're not using JSX and a build process. I completely understand the concept of just wanting to start building but we have been discussing moving all code examples to use JSX first ( see https://github.com/WordPress/gutenberg-examples/issues/171 and https://github.com/WordPress/gutenberg/issues/35776 ) and adding a new, advanced tutorial that doesn't that seems to be a bit of a step backward in that regards.

We also miss out on driving home just how powerful and versatile the `@wordpress/scripts` package is. This is a perfect opportunity to show it being used in a non-block application and would actually simplify some of the enqueue code by handling the dependencies automatically. 

From my own personal experience having built applications like this in the past, I would always choose JSX  for readability and the `@wordpress/scripts` as it provides all of the tools I would need to get starting quickly - but that's just one developer's opinion 😄

**Comment by getdave** [comment]
Is there a general tutorial on "getting setup with `@wordpress/scripts` and JSX? If there is then it would seem fine to point users towards that as a "setup" step and then continue the tutorial in JSX mode. It is far more readable and probably makes the tutorial more approachable for both beginner and more advanced readers.

File: docs/how-to-guides/data-basics/1-setup.md
Comment (getdave) [inline docs/how-to-guides/data-basics/1-setup.md]:
```suggestion
We will build the application as a WordPress plugin, which means you need to have WordPress itself installed. One way to do this is by following the instructions on the [Getting Started](/docs/contributors/code/getting-started-with-code-contribution.md) page. Once your setup is complete, you can follow along with the rest of this tutorial.
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
list of pages from the [WordPress REST API](https://developer.wordpress.org/rest-api/).
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
Before we start, let’s confirm we actually have some pages to fetch. Within WPAdmin, Navigate to Pages using the sidebar menu and
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
ensure it shows at least four or five Pages:
```

Why "positions"? I was confused as to whether you meant to use this word.

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
requested by the `getEntityRecords` resolver after first running the _selector_. If you wait a moment and re-run it, it
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
					<th>Title</th>
```

Nit.

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
## Step 4: Add a search box
```

Nit: consistent capitalisation.

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
typically solves this problem with a search box – let’s implement one too!
```
Nit: no need for comma here.

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
query parameter and uses it to  _limit results to those matching a string_. But how to use it? We can pass custom query
```

Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
query parameter and uses it to  _Limit results to those matching a string_. But how can we use it? We can pass custom query
```

File: docs/how-to-guides/data-basics/1-setup.md
Comment (adamziel) [reply docs/how-to-guides/data-basics/1-setup.md]: Thank you!

File: docs/how-to-guides/data-basics/1-setup.md
Comment (adamziel) [reply docs/how-to-guides/data-basics/1-setup.md]: Great catch!

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
Let’s mirror this in our `useSelect` call as follows:
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
The `searchTerm` is now used as a `search` query parameter when provided. Note that `searchTerm` is also specified
```

Nit: avoiding "specified" twice.

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
Finally, here’s how `MyFirstApp` looks once we wire it all together:
```

We don't need the "like" here.

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
### Using core-data instead vs calling the API directly
```

When I started reading the section I was expecting you to talk about an alternative approach that was better. I'm recommending adjusting the title to make the juxtaposition clearer.

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
Let’s take a pause for a moment to consider the downsides of an alternative approach we could have taken - working with the API directly. Imagine we sent the API requests directly:
```

Again attempting to be more explicit about the point you're trying to make here.

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
Working outside of core-data, we would need to solve two problems here.
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
Firstly, out-of-order updates. Searching for „About” would trigger five API requests filtering for `A`, `Ab`, `Abo`, `Abou`, and
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
`About`. Theese requests could finish in a different order than they started. It is possible that _search=A_ would resolve after _
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
search=About_ and thus we’d display the wrong data.
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
Secondly, every keystroke would trigger an API request. If you typed `About`, deleted it, and retyped it, it would
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
All in all, the utilities built into core-data are designed to solve the typical problems so that you can focus on your application
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
It takes the name of the selector and the _exact same arguments you passed to that selector_ and returns either `true` if the data was already loaded or `false`
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]: I'd love for to show the `getEntityRecords` selector here and extract the arguments into a variable in order that you can use it across both selector and resolver. In fact it would be even better to start out with the hard coded form and then show a refactor whereby you move to a variable for convenience. 

Why? 

1. Because it's super easy to have a resolver that doesn't "work" because you didm't pass the exact same arguments.
2. The pattern is well established in Gutenberg Core and thus pretty standard.

File: docs/how-to-guides/data-basics/README.md
Comment (getdave) [inline docs/how-to-guides/data-basics/README.md]:
```suggestion
This tutorial aims to get you comfortable with the Gutenberg data layer. It guides you through building a simple React application that enables the user to manage their WordPress pages. The finished app will look like this:
```

Aiming not to put people off 😄

**Review by getdave — COMMENTED** [review]
Amazing work Adam. The refactors you've done have made this so much better.

I re-read for accuracy and suggested some nits. 

Can't wait for this to land!

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (adamziel) [reply docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]: Such a great observation @getdave, thank you! I just introduced that refactor – would you mind reviewing if that's what you meant?

**Comment by adamziel** [comment]
Fantastic suggestions @getdave, I merged all of them – thank you so much!

I think we're almost there. I removed the compiled bits and focused on JSX as almost everyone here suggested. I will polish the code examples a bit more to use imports instead of `window.wp` and we may be good to merge.

File: docs/how-to-guides/data-basics/README.md
Comment (youknowriad) [inline docs/how-to-guides/data-basics/README.md]: I'm not sure we should keep that here.

**Review by youknowriad — COMMENTED** [review]
I see that you included the app here, and I saw that you extracted it into a PR in Gutenberg-examples as well. Should we remove the app from here?

File: docs/how-to-guides/data-basics/README.md
Comment (adamziel) [reply docs/how-to-guides/data-basics/README.md]: I removed it, thanks

**Comment by adamziel** [comment]
@youknowriad Yeah I don't like having the code in this repo AND in gutenberg-examples. I removed the samples from this PR, but there is a small downside: you can no longer review how the current step looks like when finished, only the entire tutorial. I'm considering adding `step1-setup`, `step2-list-of-pages` etc inside of the `09-code-data-basics-esnext` in gutenberg-examples, what do you think?

**Comment by getdave** [comment]
Great work on the updates Adam. Thanks for being so receptive to feedback.

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
if we’re still waiting. Let’s add it to `useSelect`:
```

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md:368]: Perfect! 🌟

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md
Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
There is just one last problem. It is easy to make a typo and pass different arguments to `getEntityRecords` and `hasFinishedResolution`. We can remove this risk by storing the arguments in a variable:
```

Oh the irony! 😉

Comment (getdave) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]:
```suggestion
There is just one last problem. It is easy to make a typo and end up passing different arguments to `getEntityRecords` and `hasFinishedResolution`. It is critical that they are identical. We can remove this risk by storing the arguments in a variable:
```

Comment (adamziel) [reply docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]: hahaha Muphry's law always works

**Comment by adamziel** [comment]
Alright, does anyone dare to approve this PR and the code sample in [gutenberg-examples](https://github.com/WordPress/gutenberg-examples/pull/191)? :D

**Comment by youknowriad** [comment]
>  I'm considering adding step1-setup, step2-list-of-pages etc inside of the 09-code-data-basics-esnext in gutenberg-examples, what do you think?

It feels a bit too much, maybe it's fine to just have these temporary steps like we have them now inline in the tutorial.

**Review by youknowriad — APPROVED** [review]
LGTM

**Comment by mkaz** [comment]
Good stuff! 👍 

For this to show up on the block editor handbook site you need to add it to the `docs/toc.json` file and then run `npm run docs:build` and check in the generated `manifest.json` file.

**Pull request merged**
